### PR TITLE
feat: backoff in retry helper

### PR DESCRIPTION
**Pull request by maan2003**
I really like 0.2, 0.2, 0.4, 0.6, 1.0, 1.6, 2.6, 4.6, 5, 5...

We should export some statergy in fedimint-core but naming is hard.

### CODE REVIEW

File: fedimint-core/Cargo.toml [hunk 1]
```diff
@@ -53,6 +53,7 @@ macro_rules_attribute = "0.2.0"
 bitvec = "1.0.1"
 parity-scale-codec = { version = "3.6.9", features = ["derive"] }
 imbl = "2.0.3"
+backon = "0.4.3"
```
Comment (dpc) [inline fedimint-core/Cargo.toml:56]: Every new dep worries me, but seems small and very popular ... https://crates.io/crates/backon...

File: fedimint-core/src/util/mod.rs [hunk 2]
```diff
@@ -320,41 +324,41 @@ pub fn handle_version_hash_command(version_hash: &str) {
 pub async fn retry<F, Fut, T>(
```
Comment (dpc) [inline fedimint-core/src/util/mod.rs:324]: Unclear if we even need it anymore. Seems like any future is retriable now? https://docs.rs/backon/latest/backon/trait.Retryable.html#tymethod.retry

File: fedimint-core/Cargo.toml (hunk 1 above)
Comment (maan2003) [reply fedimint-core/Cargo.toml:56]: maybe we just should fork it, just maths so should be pretty stable

File: fedimint-core/src/util/mod.rs (hunk 2 above)
Comment (maan2003) [reply fedimint-core/src/util/mod.rs:324]: we don't get logging, also that uses `tokio::time::sleep` (read hates wasm)

File: fedimint-core/Cargo.toml (hunk 1 above)
Comment (dpc) [reply fedimint-core/Cargo.toml:56]: Nah. We pin it, it's good. It's not a big crate.

File: fedimint-core/src/util/mod.rs (hunk 2 above)
Comment (maan2003) [reply fedimint-core/src/util/mod.rs:324]: also (|| async { }).retry(...) feels very awkward

File: fedimint-core/Cargo.toml (hunk 1 above)
Comment (maan2003) [reply fedimint-core/Cargo.toml]: pinned and added comment so someone doesn't approve dependabot :smile:

**Comment by maan2003** [comment]
adding backport label, it would be nice to have in fedi.

File: fedimint-core/src/util/mod.rs (hunk 2 above)
Comment (dpc) [reply fedimint-core/src/util/mod.rs:324]: True.

File: fedimint-core/src/util/mod.rs [hunk 3]
```diff
@@ -456,7 +466,15 @@ mod tests {
...
 
-        let _ = retry("Run 3 times".to_string(), closure, Duration::ZERO, 3).await;
+        let _ = retry(
+            "Run 3 once",
+            closure,
+            ConstantBackoff::default()
+                .with_delay(Duration::ZERO)
+                // retry two error
+                .with_max_times(2),
+        )
```
Comment (maan2003) [inline fedimint-core/src/util/mod.rs:476]: NOTE: slight change in semantics

File: fedimint-core/src/util/mod.rs [hunk 4]
```diff
@@ -320,41 +324,41 @@ pub fn handle_version_hash_command(version_hash: &str) {
 pub async fn retry<F, Fut, T>(
     op_name: impl Into<String>,
     op_fn: F,
-    interval: Duration,
-    max_attempts: u32,
+    statergy: impl backon::BackoffBuilder,
```
Comment (elsirion) [inline fedimint-core/src/util/mod.rs:327]: Could we use this opportunity to switch the argument order? It's awkward to have the closure in the middle.

File: gateway/ln-gateway/tests/integration_tests.rs [hunk 5]
```diff
@@ -1486,8 +1486,10 @@ async fn test_gateway_executes_swaps_between_connected_federations() -> anyhow::
                     };
                     Ok(post_balances)
                 },
-                Duration::from_secs(1),
-                15,
+                FibonacciBackoff::default()
+                    .with_min_delay(Duration::from_millis(200))
+                    .with_max_delay(Duration::from_secs(3))
+                    .with_max_times(10),
```
Comment (elsirion) [inline gateway/ln-gateway/tests/integration_tests.rs:1492]: New max delay: 15.6s, pretty similar to before.

```python
In [1]: def fib(start, n):
   ...:     prev = 0
   ...:     current = start
   ...:     for i in range(0,n):
   ...:         yield current
   ...:         next_current = prev + current
   ...:         prev = current
   ...:         current = next_current
   ...: 

In [2]: sum([min(3000, x) for x in fib(200, 10)])
Out[2]: 15600
```

**Review by elsirion — APPROVED** [review]

File: fedimint-core/src/util/mod.rs (hunk 4 above)
Comment (maan2003) [reply fedimint-core/src/util/mod.rs]: updated

**Review by elsirion — APPROVED** [review]

File: fedimint-core/src/util/mod.rs [hunk 6]
```diff
@@ -319,42 +323,42 @@ pub fn handle_version_hash_command(version_hash: &str) {
 ///   error of the closure is returned
 pub async fn retry<F, Fut, T>(
     op_name: impl Into<String>,
+    statergy: impl backon::BackoffBuilder,
```
Comment (joschisan) [inline fedimint-core/src/util/mod.rs:326]: typo

**Review by joschisan — APPROVED** [review]

File: fedimint-core/src/util/mod.rs (hunk 6 above)
Comment (maan2003) [reply fedimint-core/src/util/mod.rs:326]: we can live with typo :)
much better than reacquiring approvals

**Comment by fedimint-backports** [comment]
Successfully created backport PR for `releases/v0.3`:
- #4769

**Pull request merged**